pkg/remotecmd/kubeexec.go

```diff
@@ -0,0 +1,294 @@
+func (k *kubeExecProcessHandler) StartProcessForCommand(
+	devfileCmd devfilev1.Command,
+	kclient kclient.ClientInterface,
+	podName string,
+	containerName string,
+	outputHandler CommandOutputHandler,
+) error {
+	klog.V(4).Infof("StartProcessForCommand for %q", devfileCmd.Id)
+
+	if devfileCmd.Exec == nil {
+		return errors.New(" only Exec commands are supported")
+	}
+
+	// deal with environment variables
+	cmdLine := devfileCmd.Exec.CommandLine
+	setEnvVariable := util.GetCommandStringFromEnvs(devfileCmd.Exec.Env)
+	if setEnvVariable != "" {
+		cmdLine = setEnvVariable + " && " + devfileCmd.Exec.CommandLine
+	}
+
+	// Change to the workdir and execute the command
+	pidFile := getPidFileForCommand(devfileCmd)
+	cmd := []string{common.ShellExecutable, "-c"}
+	// Storing the /bin/sh parent process PID. It will allow to determine its children later on and kill them when a stop is request
+	pidWriterCmd := fmt.Sprintf("echo $$ > %s", pidFile)
+	// Redirecting to /proc/1/fd/* allows to redirect the process output to the output streams of PID 1 process inside the container.
+	// This way, returning the container logs with 'odo logs' or 'kubectl logs' would work seamlessly.
+	// See https://stackoverflow.com/questions/58716574/where-exactly-do-the-logs-of-kubernetes-pods-come-from-at-the-container-level
+	outputRedirectCmd := "1>>/proc/1/fd/1 2>>/proc/1/fd/2"
+	if devfileCmd.Exec.WorkingDir != "" {
+		// since we are using /bin/sh -c, the command needs to be within a single double quote instance,
+		// for example "cd /tmp && pwd"
+		// Full command is: /bin/sh -c "echo $$ > $pidFile && cd $workingDir && ($cmdLine) 1>>/proc/1/fd/1 2>>/proc/1/fd/2"
+		// We are deleting the PID file if the main command does not succeed, so its status could be detected accordingly.
+		cmd = append(cmd, fmt.Sprintf("%s && cd %s && (%s) %s",
+			pidWriterCmd, devfileCmd.Exec.WorkingDir, cmdLine, outputRedirectCmd))
+	} else {
+		// Full command is: /bin/sh -c "echo $$ > $pidFile && ($cmdLine) 1>>/proc/1/fd/1 2>>/proc/1/fd/2"
+		// We are deleting the PID file if the main command does not succeed, so its status could be detected accordingly.
+		cmd = append(cmd, fmt.Sprintf("%s && (%s) %s", pidWriterCmd, cmdLine, outputRedirectCmd))
+	}
+
+	go func() {
+		if outputHandler != nil {
+			outputHandler(Starting, nil, nil, nil)
+		}
+
+		stdout, stderr, err := ExecuteCommandAndGetOutput(kclient, podName, containerName, false, cmd...)
+		if err != nil {
+			klog.V(2).Infof("error while running background command: %v", err)
+		}
+
+		if outputHandler != nil {
+			outputHandler(Stopped, stdout, stderr, err)
+		}
+	}()
+
+	return nil
+}
+
+// StopProcessForCommand stops the process representing the specified Devfile command.
+// Because of the way this process is launched and its PID stored (see StartProcessForCommand),
+// we need to determine the process children (there should be only one child). Then killing those children
```

Comment:
Is there a reason why there should only be one child?

Reply:
The Devfile command is executed in a subshell, which is the child here (this is the intent of the `($cmdLine)` you asked about previously).
This was done to make sure we could handle cases where the command-line string defined in the Devfile does more things, like 'sleep 10; <do_something_in_background> & <start_my_server>'